### PR TITLE
feat(mempool): Add new Dynamic Optimal Graph (DOG) gossip protocol

### LINKED ISSUE #4558: mempool/dog: Implementation
Implement the DOG protocol following the [spec](https://github.com/cometbft/cometbft/blob/main/spec/mempool/gossip/dog.md) and [ADR](https://github.com/cometbft/cometbft/pull/4493) .

### PR DESCRIPTION
**Pull request by hvanz**
Closes #4558 

The bulk of the work is in one file: `mempool/reactor.go`.

Each commit is a self-contained addition to the code:
- [add new proto messages HaveTx and ResetRoute](https://github.com/cometbft/cometbft/pull/4555/commits/e2364279afb3973c0cf1d0908c30c288c10664a1)
- [add config](https://github.com/cometbft/cometbft/pull/4555/commits/bca343a201b6bfc514b502d82a75a86396677442)
- [add MempoolControlChannel](https://github.com/cometbft/cometbft/pull/4555/commits/cb65e1556aadb102d8cad6de6c1929744529dad1)
- [add GetSenders method to Mempool interface, and Senders to Entry interface](https://github.com/cometbft/cometbft/pull/4555/commits/d73f263a7fde1191b54e6d81445aa70e19d179fc)
- [add router to mempool reactor](https://github.com/cometbft/cometbft/pull/4555/commits/fb38f14a10752a63213a2211b4cb4399829a9a0f)
- [add redundancy controller to mempool reactor](https://github.com/cometbft/cometbft/pull/4555/commits/f91be4355a797fc81c11f7d939171af53fb0d401)
- [add metrics DisabledRoutes and Redundancy](https://github.com/cometbft/cometbft/pull/4555/commits/28b14f199ad7ca7b0f5c34567983e01f2ecb53f3)
- [add changelog file](https://github.com/cometbft/cometbft/pull/4555/commits/4677722b30167e24f54a5b3a6e208aeaf1d83eb3)

---

#### PR checklist

- [X] Tests written/updated
- [X] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [X] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: .changelog/unreleased/features/3297-mempool-interface-senders.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- `[mempool]` Add new `GetSenders` method to `Mempool` interface, and `Senders` method to `Entry`
```
Comment (jmalicevic) [inline .changelog/unreleased/features/3297-mempool-interface-senders.md:1]: This has to be duplicated under breaking change too.

Comment (hvanz) [reply .changelog/unreleased/features/3297-mempool-interface-senders.md]: I moved the file to breaking-changes.

File: api/cometbft/mempool/v1/message.go [hunk 2]
```diff
@@ -13,12 +13,28 @@ func (m *Txs) Wrap() proto.Message {
 	return mm
 }
 
+func (m *HaveTx) Wrap() proto.Message {
```
Comment (melekes) [inline api/cometbft/mempool/v1/message.go:16]: Is the plan to have DOG in `v1`?

File: mempool/reactor.go [hunk 3]
```diff
@@ -12,12 +12,19 @@ import (
...
+	cmtsync "github.com/cometbft/cometbft/libs/sync"
 	"github.com/cometbft/cometbft/p2p"
 	tcpconn "github.com/cometbft/cometbft/p2p/transport/tcp/conn"
 	"github.com/cometbft/cometbft/types"
 )
 
+// A number in the open interval (0, 100) representing a percentage of
+// config.DOGTargetRedundancy. In the DOG protocol, it defines acceptable lower
+// and upper bounds for redundancy levels as a deviation from the target value.
+const TargetRedundancyDeltaPercent = 10
```
Comment (melekes) [inline mempool/reactor.go:26]: Why it's public?

File: mempool/reactor.go [hunk 4]
```diff
@@ -134,31 +161,87 @@ func (memR *Reactor) AddPeer(peer p2p.Peer) {
...
-			return
+	senderID := e.Src.ID()
+
+	switch e.ChannelID {
+	case MempoolControlChannel:
+		switch msg := e.Message.(type) {
+		case *protomem.HaveTx:
+			txKey := types.TxKey(msg.GetTxKey())
+			if len(txKey) == 0 {
+				memR.Logger.Error("Received empty HaveTx message from peer", "src", e.Src.ID())
```
Comment (melekes) [inline mempool/reactor.go:190]: IMHO, this should also lead to us disconnecting from this peer, but that's a point for another PR, I guess.

File: mempool/reactor.go [hunk 5]
```diff
@@ -317,3 +434,216 @@ func (memR *Reactor) broadcastTxRoutine(peer p2p.Peer) {
...
+	sourcesWithTarget := make([]p2p.ID, 0)
+	for s, targets := range r.disabledRoutes {
+		if _, ok := targets[target]; ok {
+			sourcesWithTarget = append(sourcesWithTarget, s)
+		}
+	}
+
+	if len(sourcesWithTarget) > 0 {
+		randomSource := sourcesWithTarget[cmtrand.Intn(len(sourcesWithTarget))]
+		delete(r.disabledRoutes[randomSource], target)
```
Comment (melekes) [inline mempool/reactor.go:511]: To check. Shouldn't we call `resetRoutes(randomSource)` here?

File: mempool/reactor.go [hunk 6]
```diff
@@ -317,3 +434,216 @@ func (memR *Reactor) broadcastTxRoutine(peer p2p.Peer) {
...
+	}
+}
+
+// numRoutes returns the number of disabled routes in this node. Used for
+// metrics.
+func (r *gossipRouter) numRoutes() int {
+	r.mtx.RLock()
```
Comment (melekes) [inline mempool/reactor.go:518]: How often does this get accessed? Maybe it makes sense to change this to atomic var?

**Review by melekes — COMMENTED** [review]
great work @hvanz and @jmalicevic  👍

File: api/cometbft/mempool/v1/message.go (hunk 2 above)
Comment (hvanz) [reply api/cometbft/mempool/v1/message.go:16]: Not when we are so close to release `v1` but it's possible that will land in `v1.x-experimental`.

File: mempool/reactor.go (hunk 4 above)
Comment (hvanz) [reply mempool/reactor.go:190]: That's correct. For another PR.

File: mempool/reactor.go (hunk 3 above)
Comment (hvanz) [reply mempool/reactor.go]: True. I made it private.

File: mempool/reactor.go (hunk 5 above)
Comment (hvanz) [reply mempool/reactor.go:511]: No. `resetRoutes(peerID)` removes all routes with `peerID` as source or target. Here we just want to remove one route.

File: mempool/reactor.go (hunk 6 above)
Comment (hvanz) [reply mempool/reactor.go:518]: The function `numRoutes` is only used for metrics but it reads `disabledRoutes`, which is used quite often. I don't think it makes sense to make it an atomic var.

File: api/cometbft/mempool/v1/message.go (hunk 2 above)
Comment (melekes) [reply api/cometbft/mempool/v1/message.go:16]: Then maybe we should create proto `v2`?

Comment (hvanz) [reply api/cometbft/mempool/v1/message.go:16]: There's definitely material for `v2` in `main`.

Comment (melekes) [reply api/cometbft/mempool/v1/message.go:16]: Should we do it here or open another PR?

Comment (hvanz) [reply api/cometbft/mempool/v1/message.go:16]: Sorry, I misunderstood the original question. I thought you were asking about Comet's version. 

I agree we should update the mempool proto version to `v2`. Will do it here in this PR.

Comment (hvanz) [reply api/cometbft/mempool/v1/message.go]: Done.

**Review by jmalicevic — APPROVED** [review]

**Review by melekes — APPROVED** [review]

**Pull request merged**